### PR TITLE
Small improvements

### PR DESCRIPTION
**Pull request by ricardomaraschini**
Well, you should go commit by commit if you want to understand what has changed here, a simple list:

1. Fixed problem when running snitch outside of the project directory
2. Added support file strange file names containing "."
3. Fix unable to see multiple tests due to clear.
4. Deduplicating tests for changes on multiple files on the same dir.
5. Removed uneeded select.
6. Renaming variable to avoid shadowing with a global.
7. Solved possible panic() by properly validating Stat() return.
8. Properly validating command line arguments.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
 	rootPath := flag.String("path", defaultPath, "the root path to be watched")
-	interval := flag.Int("interval", 5, "the interval (in seconds) for scanning files")
+	interval := flag.Duration("interval", 5*time.Second, "the interval (in seconds) for scanning files")
 	flag.Parse()
 
-	watchedFiles = walk(rootPath)
-	scheduleScanAt(rootPath, interval)
-}
+	if *interval < 0 {
+		log.Fatal("negative interval:", *interval)
```
Comment (axcdnt) [inline main.go]:
```suggestion
		log.Fatal("invalid interval, must be > 0:", *interval)
```

File: main.go [hunk 2]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
-func scheduleScanAt(rootPath *string, interval *int) {
-	runTicker := time.NewTicker(
-		time.Duration(time.Duration(*interval) * time.Second))
-	for {
-		select {
-		case <-runTicker.C:
-			scan(rootPath)
-		}
+	if err := os.Chdir(*rootPath); err != nil {
+		log.Fatal("Chdir:", err)
```
Comment (axcdnt) [inline main.go]:
```suggestion
		log.Fatal("could not change directory:", err)
```

File: main.go [hunk 3]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
 
-func scheduleScanAt(rootPath *string, interval *int) {
-	runTicker := time.NewTicker(
-		time.Duration(time.Duration(*interval) * time.Second))
-	for {
-		select {
-		case <-runTicker.C:
-			scan(rootPath)
-		}
+	if err := os.Chdir(*rootPath); err != nil {
```
Comment (axcdnt) [inline main.go:34]: 👍

File: main.go [hunk 4]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
+	}
+
+	watchedFiles = walk(rootPath)
+	for range time.NewTicker(*interval).C {
+		scan(rootPath)
 	}
 }
 
 func scan(rootPath *string) {
+	changedDirs := make(map[string]bool, 0)
```
Comment (axcdnt) [inline main.go]: What do you think of changing `changed` to `modified`, so it stays more consistent with the `mod` of Go's file API?

```suggestion
	modifiedDirs := make(map[string]bool, 0)
```

File: main.go [hunk 5]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
 }
 
 func walk(rootPath *string) FileInfo {
-	watchedFiles := FileInfo{}
-	err := filepath.Walk(*rootPath, visit(watchedFiles))
-	if err != nil {
-		panic(err)
+	wf := FileInfo{}
+	if err := filepath.Walk(*rootPath, visit(wf)); err != nil {
+		log.Fatal("Walk:", err)
```
Comment (axcdnt) [inline main.go]:
```suggestion
		log.Fatal("could not traverse files:", err)
```

File: main.go [hunk 6]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
...
+	if isTestFile(filePath) {
+		// change happened on a test file, we need to test.
+		return true
 	}
-}
 
-func findTestFilePath(filePath string) string {
+	// changed file is not a test file, we return true only if there is a
+	// test file for the go file.
 	fileName := path.Base(filePath)
```
Comment (axcdnt) [inline main.go]: I'd turn this logic into a new function like `func hasTestFile(filePath string) bool`. In this case, this function stays more readable like:

```
func needsTest(filePath string) bool {
	if isTestFile(filePath) || hasTestFile(filePath) {
		return true
	}

	return false
}
```

File: main.go [hunk 7]
```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
 var watchedFiles = FileInfo{}
 
 func main() {
-	defaultPath, _ := os.Getwd()
+	defaultPath, err := os.Getwd()
+	if err != nil {
+		log.Fatal("Getwd: ", err)
```
Comment (axcdnt) [inline main.go]:
```suggestion
		log.Fatal("could not get current directory: ", err)
```

**Review by axcdnt — CHANGES_REQUESTED** [review]
@ricardomaraschini great PR in general! Just found minor details and a few readability improvements.

File: main.go [hunk 8]
```diff
@@ -19,105 +18,119 @@ type FileInfo map[string]time.Time
...
+	}
+
+	watchedFiles = walk(rootPath)
+	for range time.NewTicker(*interval).C {
+		scan(rootPath)
 	}
 }
 
 func scan(rootPath *string) {
+	modifiedDirs := make(map[string]bool, 0)
```
Comment (axcdnt) [inline main.go:45]:
```suggestion
	modifiedDirs := []string{}
```

File: main.go [hunk 9]
```diff
@@ -19,105 +18,119 @@ type FileInfo map[string]time.Time
...
+			continue
+		}
+
+		watchedFiles[filePath] = mostRecentModTime
+		if !needsTest(filePath) {
+			continue
 		}
+
+		pkgDir := path.Dir(filePath)
+		modifiedDirs[pkgDir] = true
```
Comment (axcdnt) [inline main.go:70]:
```suggestion
		modifiedDirs = append(modifiedDirs, pkgDir)
```

File: main.go [hunk 10]
```diff
@@ -19,105 +18,119 @@ type FileInfo map[string]time.Time
...
+	}
+
+	if len(modifiedDirs) == 0 {
+		return
 	}
+
+	dedup := make([]string, 0)
```
Comment (axcdnt) [inline main.go:77]: By changing the data structure of `modifiedDirs` from `map` to `slice`, I think this `dedup` logic can be removed. You can call `test(modifiedDirs)` directly.

**Review by axcdnt — CHANGES_REQUESTED** [review]

File: main.go (hunk 8 above)
Comment (ricardomaraschini) [reply main.go:45]: You clearly did not understand the code.

File: main.go (hunk 9 above)
Comment (ricardomaraschini) [reply main.go:70]: see my previous comment.

**Comment by ricardomaraschini** [comment]
I am closing this PR as it is becoming bikeshedding.

**Pull request closed without merging**